Use content hashes to invalidate assets instead of last modified timestamps

jakemac53 opened the pull request:
All `AssetNode`s now have a `Digest` field (from package:crypto). These are computed eagerly, and must be available for any node which has a corresponding asset that exists. It is also serialized as a part of the asset graph (base64 encoded for json compatibility for now).

There is now an additional interface `DigestAssetReader` which adds the `Future<Digest> digest(id)` method to a normal asset reader, and the `RunnerAssetReader` interface now requires that interface to be implemented.

`AssetGraph.build` is now a static method which returns a `Future<AssetGraph>`, and requires a `DigestAssetReader`. It eagerly reads in digests for all source assets.

MD5 is the current hash function used, because collisions are not a concern in this case (we don't use the hashes for lookups, only to compare new/old values for a single file). Bazel also uses md5 today, as well as the analyzer for summaries, so this is also consistent with those (although they may change it is still kind of nice). It is also significantly faster than sha1 in practice.

This does complicate some of the tests a bit unfortunately (search this pr for `computeHash` in tests....), but its a necessary evil imo. I don't see a way to simplify it significantly.

I didn't yet completely remove the `lastModified` method, it is still used when checking if a build script itself has been updated. I wanted to keep the scope of this as small as I could, so I will follow up with a different pull request to update that as well, and remove `lastModified` entirely.

Partially fixes https://github.com/dart-lang/build/issues/536
Unblocks https://github.com/dart-lang/build/issues/371 and https://github.com/dart-lang/build/issues/412

jakemac53 commented:
@natebosch after thinking a bit more on this I think it might make sense to change this field on `AssetNode` to be a `Future<Digest>`, and compute it lazily for at least some assets (specifically for assets that have no outputs).

The main reasoning be that you might be serving large assets like images, and we likely don't want to be hashing those on startup.

natebosch commented:
I'm very in favor of setting the interface to `Future` to allow laziness

natebosch commented on build_runner/lib/src/asset/reader.dart:
```diff
@@ -2,18 +2,34 @@ import 'dart:async';
 typedef Future RunPhaseForInput(int phaseNumber, AssetId primaryInput);
 
+abstract class DigestAssetReader implements AssetReader {
+  /// Asynchronously compute the digest for [id].
```
should we document expected behavior of this call if the Asset does not exist?

jakemac53 replied:
done

jakemac53 commented:
> Should we have any new tests for this?

I will at least add some unit tests for the MD5DigestReader, but for the most part we don't need any additional tests I don't think (already updated the asset graph tests, and the e2e tests aren't affected).

natebosch commented:
> but for the most part we don't need any additional tests I don't think

Do we have a test for the case in #536 ?

jakemac53 commented:
Also just fyi in terms of numbers, for angular_components_example:

- Creating a new asset graph takes ~1 second, doesn't noticeably affect the first build time
- Updating asset graph on incremental startup takes ~750ms
- Ram usage isn't noticeably affected

So, actually surprisingly little negative affect for angular_components_example. As mentioned earlier I could imagine a case where this would try to compute hashes of very large files which would not be good, but we can address that by making the hash be computed lazily for nodes with no output.

jakemac53 commented:
> Do we have a test for the case in #536 ?

Don't think so (otherwise it would be failing), I can add something.

jakemac53 commented:
@natebosch just fyi after playing around with a few options I don't think I want to go the `Future<Digest>` route, because that means every `AssetNode` needs a reference to a `DigestReader`.

Instead I changed the field from `digest` to `lastKnownDigest`, and changed the docs to say it may be null if the node has no outputs. As indicated this field will now only eagerly be computed for nodes that actually have outputs. It will be lazily computed and assigned if the asset is read by any build step as well (to ensure its consistent with the state in which it was last read).

natebosch commented:
SGTM

jakemac53 commented:
Actually - this doesn't fully fix https://github.com/dart-lang/build/issues/536 because the last modified logic is still used for build script update detection.

Given that I don't think its worth adding that test as a part of this PR (any reasonable one would have to be skipped anyways).

jakemac53 commented on build_runner/lib/src/asset/cache.dart:
```diff
@@ -36,6 +38,9 @@ class CachingAssetReader implements AssetReader {
       _canReadCache.putIfAbsent(id, () => _delegate.canRead(id));
 
   @override
+  Future<Digest> digest(id) => _delegate.digest(id);
```
Note that I removed caching of digests from here - they are cached on `AssetNode`s already so this would be duplicative.

jakemac53 commented:
Ok @natebosch this is actually ready for a review now :)

natebosch commented on build_runner/lib/src/asset_graph/serialization.dart:
```diff
@@ -121,6 +131,28 @@ class _AssetGraphSerializer {
 }
 
+/// Used to serialize the type of a node using an int.
+enum _NodeTypeId { Source, Synthetic, Generated }
```
[nit] I'm not sure 'Id' is useful in these enum names

jakemac53 replied:
done
